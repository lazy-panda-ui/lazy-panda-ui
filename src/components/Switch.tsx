import React from 'react';
import {
  Switch as RNSwitch,
  SwitchProps as RNSwitchProps,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';
import { useTheme } from '../theme/ThemeProvider';

export type SwitchSize = 'small' | 'medium' | 'large';
export type SwitchVariant = 'default' | 'tonal' | 'outline';

export interface SwitchProps extends Omit<RNSwitchProps, 'trackColor' | 'thumbColor' | 'ios_backgroundColor'> {
  /**
   * Size of the switch
   * @default 'medium'
   */
  size?: SwitchSize;
  /**
   * Visual variant of the switch
   * @default 'default'
   */
  variant?: SwitchVariant;
  /**
   * Color for the switch when on. Defaults to theme.colors.primary
   */
  activeColor?: string;
  /**
   * Color for the track when off. Defaults to theme.colors.outline
   */
  inactiveColor?: string;
  /**
   * Additional styles for the container
   */
  style?: ViewStyle;
  /**
   * Test ID for testing
   */
  testID?: string;
}

export const Switch: React.FC<SwitchProps> = ({
  size = 'medium',
  variant = 'default',
  activeColor,
  inactiveColor,
  style,
  disabled,
  value,
  onValueChange,
  testID,
  ...rest
}) => {
  const theme = useTheme();

  const getTrackColor = (isActive: boolean) => {
    if (disabled) return theme.colors.disabled;
    
    if (isActive) {
      switch (variant) {
        case 'tonal':
          return activeColor || theme.colors.primaryContainer;
        case 'outline':
          return 'transparent';
        default:
          return activeColor || theme.colors.primary;
      }
    } else {
      switch (variant) {
        case 'tonal':
          return inactiveColor || theme.colors.surfaceVariant;
        case 'outline':
          return 'transparent';
        default:
          return inactiveColor || theme.colors.outline;
      }
    }
  };

  const getThumbColor = (isActive: boolean) => {
    if (disabled) return theme.colors.onDisabled;
    
    if (isActive) {
      switch (variant) {
        case 'tonal':
          return activeColor || theme.colors.onPrimaryContainer;
        case 'outline':
          return activeColor || theme.colors.primary;
        default:
          return theme.colors.onPrimary;
      }
    } else {
      return theme.colors.onSurface;
    }
  };

  const getSizeStyles = (size: SwitchSize): ViewStyle => {
    const scales = {
      small: 0.8,
      medium: 1,
      large: 1.2,
    };
    const scale = scales[size];
    
    return {
      transform: [{ scale }],
    };
  };

  const styles = StyleSheet.create({
    container: {
      opacity: disabled ? 0.6 : 1,
      ...getSizeStyles(size),
    },
    outline: {
      ...(variant === 'outline' && {
        borderWidth: 2,
        borderColor: value ? (activeColor || theme.colors.primary) : theme.colors.outline,
        borderRadius: 16,
        padding: 2,
      }),
    },
  });

  return (
    <View style={[styles.container, styles.outline, style]} testID={testID}>
      <RNSwitch
        value={value}
        onValueChange={disabled ? undefined : onValueChange}
        trackColor={{
          false: getTrackColor(false),
          true: getTrackColor(true),
        }}
        thumbColor={getThumbColor(value || false)}
        ios_backgroundColor={getTrackColor(false)}
        disabled={disabled}
        {...rest}
      />
    </View>
  );
};
