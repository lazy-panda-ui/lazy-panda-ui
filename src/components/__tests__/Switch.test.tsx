import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Switch } from '../Switch';

describe('Switch', () => {
  it('renders with value', () => {
    const { toJSON } = render(<Switch value={true} onValueChange={() => {}} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onValueChange', () => {
    const onValueChange = jest.fn();
    const { getByTestId } = render(
      <Switch value={false} onValueChange={onValueChange} testID="switch" />
    );
    fireEvent(getByTestId('switch'), 'valueChange', { value: true });
    expect(onValueChange).toHaveBeenCalledWith({ value: true });
  });
});
