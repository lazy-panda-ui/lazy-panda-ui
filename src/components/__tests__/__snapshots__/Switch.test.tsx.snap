// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders with value 1`] = `
<RCTSwitch
  accessibilityRole="switch"
  onChange={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onTintColor="#007AFF"
  style={
    [
      {
        "height": 31,
        "width": 51,
      },
      {
        "backgroundColor": "#E5E5EA",
        "borderRadius": 16,
      },
    ]
  }
  thumbTintColor="#007AFF"
  tintColor="#E5E5EA"
  value={true}
/>
`;
